src/core/linux/X11.zig

```diff
@@ -2,24 +2,823 @@ const Linux = @import("../Linux.zig");
+        .libxrr = libxrr,
+        .empty_event_pipe = try std.posix.pipe(),
+        .gl_ctx = null,
+        .width = window_attrs.width,
+        .height = window_attrs.height,
+        .wm_protocols = libx11.XInternAtom(display, "WM_PROTOCOLS", c.False),
+        .wm_delete_window = libx11.XInternAtom(display, "WM_DELETE_WINDOW", c.False),
+        .net_wm_ping = libx11.XInternAtom(display, "NET_WM_PING", c.False),
+        .net_wm_state_fullscreen = libx11.XInternAtom(display, "_NET_WM_STATE_FULLSCREEN", c.False),
+        .net_wm_state = libx11.XInternAtom(display, "_NET_WM_STATE", c.False),
+        .net_wm_state_above = libx11.XInternAtom(display, "_NET_WM_STATE_ABOVE", c.False),
+        .net_wm_window_type = libx11.XInternAtom(display, "_NET_WM_WINDOW_TYPE", c.False),
+        .net_wm_window_type_dock = libx11.XInternAtom(display, "_NET_WM_WINDOW_TYPE_DOCK", c.False),
+        .net_wm_bypass_compositor = libx11.XInternAtom(display, "_NET_WM_BYPASS_COMPOSITOR", c.False),
+        .motif_wm_hints = libx11.XInternAtom(display, "_MOTIF_WM_HINTS", c.False),
+        .root_window = root_window,
+        .window = window,
+        .hidden_cursor = libx11.XCreatePixmapCursor(display, blank_pixmap, blank_pixmap, &color, &color, 0, 0),
+        .backend_type = try Core.detectBackendType(options.allocator),
+        .refresh_rate = refresh_rate,
+        .title = options.title,
+        .display_mode = .windowed,
+        .border = options.border,
+        .headless = options.headless,
+        .size = window_size,
+        .cursors = std.mem.zeroes([@typeInfo(CursorShape).Enum.fields.len]?c.Cursor),
+        .surface_descriptor = surface_descriptor,
+        .libxkbcommon = try LibXkbCommon.load(),
+    };
+    _ = libx11.XSetErrorHandler(errorHandler);
+    _ = libx11.XInitThreads();
+    _ = libx11.XrmInitialize();
+    defer _ = libx11.XFreeColormap(display, colormap);
+    for (0..2) |i| {
+        const sf = try std.posix.fcntl(x11.empty_event_pipe[i], std.posix.F.GETFL, 0);
+        const df = try std.posix.fcntl(x11.empty_event_pipe[i], std.posix.F.GETFD, 0);
+        _ = try std.posix.fcntl(x11.empty_event_pipe[i], std.posix.F.SETFL, sf | @as(u32, @bitCast(std.posix.O{ .NONBLOCK = true })));
+        _ = try std.posix.fcntl(x11.empty_event_pipe[i], std.posix.F.SETFD, df | std.posix.FD_CLOEXEC);
+    }
+    var protocols = [_]c.Atom{ x11.wm_delete_window, x11.net_wm_ping };
+    _ = libx11.XSetWMProtocols(x11.display, x11.window, &protocols, protocols.len);
+    _ = libx11.XStoreName(x11.display, x11.window, options.title.ptr);
+    _ = libx11.XSelectInput(x11.display, x11.window, set_window_attrs.event_mask);
+    _ = libx11.XMapWindow(x11.display, x11.window);
+    _ = libx11.XGetWindowAttributes(x11.display, x11.window, &window_attrs);
+    const backend_type = try Core.detectBackendType(options.allocator);
+    switch (backend_type) {
+        .opengl, .opengles => {
+            if (libgl != null) {
+                // zig fmt: off
```

Comment:
nit: prefer to never have zig fmt directives, even if it produces worse formatting

Reply:
Noted for my next X11 PR!